mithril-common/src/chain_observer/pallas_observer.rs

```diff
@@ -239,6 +234,65 @@ impl PallasChainObserver {
         Ok(Some(stake_distribution))
     }
 
+    /// Calculates the current KES period
+    /// using the provided `chain_point`and `slots_per_kes_period`.
+    async fn calculate_kes_period(
+        &self,
+        chain_point: Point,
+        slots_per_kes_period: u64,
+    ) -> Result<u32, ChainObserverError> {
+        match slots_per_kes_period {
+            slots if slots > 0 => {
+                let current_kes_period = chain_point.slot_or_default() / slots;
+                Ok(current_kes_period as u32)
+            }
+            _ => Err(anyhow!(
+                "PallasChainObserver failed to calculate kes period"
+            ))
+            .with_context(|| "slots_per_kes_period must be greater than 0")?,
+        }
+    }
```

Comment:
The "match if" here make the code hard to read, an early return would be better.
Also the error context, `PallasChainObserver failed to calculate kes period`, should be added by the caller `get_key_period`.

_Note: I changed the error type for this method to allow to use the `anyhow::bail` macro._

```suggestion
    async fn calculate_kes_period(
        &self,
        chain_point: Point,
        slots_per_kes_period: u64,
    ) -> Result<u32, StdError> {
        if slots_per_kes_period == 0 {
            anyhow::bail!("slots_per_kes_period must be greater than 0");
        }

        let current_kes_period = chain_point.slot_or_default() / slots_per_kes_period;
        Ok(current_kes_period as u32)
    }
```

Reply:
thank you,

updated but kept the pattern we are following to handle and trace errors adding more troubleshoot context using:

```rust
async fn calculate_kes_period(
    &self,
    chain_point: Point,
    slots_per_kes_period: u64,
) -> Result<KESPeriod, ChainObserverError> {
    if slots_per_kes_period == 0 {
        return Err(anyhow!("slots_per_kes_period must be greater than 0"))
            .with_context(|| "PallasChainObserver failed to calculate kes period")?;
    }

    let current_kes_period = chain_point.slot_or_default() / slots_per_kes_period;
    Ok(u32::try_from(current_kes_period)
        .map_err(|err| anyhow!(err))
        .with_context(|| "PallasChainObserver failed to convert kes period")?)
}
```